rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,129 @@
+
+After the acknowledgement, both sides will send a count of how many bytes they have received in the connection.
+
+### Data exchange
+
+After both sides agree that the connection can proceed, data is exchanged in frames consisting of an integer count of how many bytes are being acknowledged as successfully received (since the previous acknowledgement, or since the beginning of the connection, to reduce the size of the variable-length integer), then a length-prefixed chunk of data. The size of the chunk must not exceed a size of 128KiB. If either side cannot proceed from the exchanged resume point, or if either side wants to explicitly close the connection, a value of `0xffff_ffff_ffff_ffff` should be sent (as the acknowledgement). Upon receiving such value, the connection should be considered closed by the remote side.
+
+### State
+
+Both sides need to maintain a log of data that has been generated by the application and potentially sent over the wire, but hasn't yet been acknowledged by the peer. Such a buffer should be limited in size to avoid unconstrained resource consumption - from some early tests, a buffer size of 2MiB seems to be sufficient to maintain near full bandwidth when transferring data across a single SSH channel, compared to a regular SSH connection that doesn't make use of the resumption protocol. The protocol does not mandate the use of a buffer with a fixed maximum size, so other approaches can be used in the future if need be.
+
+This design allows for keepalive frames containing no acknowledgement and no data, similarly to the protocol used in [TLS routing Ping](https://github.com/gravitational/teleport/blob/master/rfd/0081-tls-ping.md); a frame consisting of two NULs should be sent if no data has been otherwise sent for some _keepalive interval_ (30 seconds?), and the underlying connection should be closed and considered dead if no data has been received for two or three times that interval.
+
+The server side of the connection is tasked with keeping track of the existing resumable connections, and should take care not to let dead connections accumulate if they're unused. A connection that has no underlying transport for longer than a _grace period_ - we have an idle timeout of 15 minutes for SSH connections, but we can pick something slightly shorter, around 5 minutes - should be closed and then cleaned up by the server. Closed connections (as a result of a SSH error, or as a result of the peer sending a close signal) should likewise be cleaned up.
```

Comment:
In addition to the timeout we should also consider connection limits.  I don't know if we have a configurable global connection limit?  But if not I suggest we add one with a sane default.

Then we should ensure that inactive + active connection counts remain under that configured limit.  If too many active connections exist, then LRU removal of inactive sessions seems to make sense to me.

Reply:
Since we're already adding server-side application-layer support (to enable connection roaming) maybe we can have some tight limits (including for buffer sizes) for non-authenticated connections, up to the point where non-authenticated connections just can't be resumed if there's too many open connections already. I'd be really really wary of affecting authenticated connections tho, since we don't really have a grasp on how many concurrent sessions can be occasionally expected.

Reply:
I can understand the concern around global limits.  But at a minimum I think per-user limits should also make sure that resumable sessions are also considered as part of their limit (and removed in a LRU if the limit is hit).

Reply:
Why are the resumable connections special in this regard? We don't seem to have a global limit with a LRU cache for SSH connections over a standard connection.

Reply:
Sorry I was not very clear.

I think a global limit for connections makes sense generally as part of our DoS defense strategy.  But I recognize that's outside of the scope of this RFD.

That said, I think it's important the that the per-user limit that we do have today also considers not just active connections but inactive connections that can be resumed.  I think this is important because if allow only a time base limit, this opens up a new DoS vector.  One which does not even require the connection to be active to impact the resources of the service.

So I just want to make sure that we consider these inactive sessions part of that existing per-user limit.  Let me know if you have any more thoughts or questions, thank you!

Reply:
I reworded some of the resource exhaustion/limit details in e6c9c157df2eacd760f36ba5f6237b80b44442d7, PTAL. It's not very prescriptive, but a lot of these details can only be fully defined during the implementation phase, I feel.